0030-simple-classical-representations.md

```diff
@@ -0,0 +1,260 @@
+These expressions will not be valid in the general `Instruction.condition` field; we do not want to expand any support for something we are already moving to remove.
+
+
+#### Stage 2
+
+Looking beyond the initial phase, it will be convenient to define a less verbose way for users to create these expressions.
+Since this document is intended to define a representation that can be removed and deprecated from Terra easily if it needs to evolve, we will _not_ consider overloading the Python magic-methods `__and__` etc on `Clbit` and `ClassicalRegister` in general Terra usage.
+However, we can do so in a limited scope, which will temporarily change the semantics.
+This will look like (continued from above code block, and the same examples in the same order):
+
+```python
+with expr.build():
+    equal_reg = cr1 == 2
+    equal_bit = loose[0] == False
+    not_bit = ~loose[0]
+    bounded_reg = (0 < cr1) & (cr1 <= cr2)
+```
+
+It is not possible to overload the behaviour of the Python Boolean operators (`not`, `and` and `or`) for classes, which is why the bitwise operations `~` and `&` are used instead (with the corresponding precedence frustrations).
+Within the `expr.build()` context, the magic methods of `ClassicalRegister` and `Clbit` will be temporarily overridden to be an eager builder interface for the syntax tree given above.
+This means that the expressions will not need to be bound to variables in order to use them, as the resulting objects will be well-defined and the correct tree form.
+For example, it will be valid to do
+
+```python
+qc = QuantumCircuit([Qubit()], loose)
+with expr.build():
+    with qc.if_test(~loose[0]):
+        qc.h(0)
+```
+
+The reason to require entering the builder context to make the magic methods available is for potential future ease of change/removal/deprecation.
+The mutation of global state required to do this makes the builder interface non-thread-safe, but this is not judged to be a problem; it is uncommon to attempt to multithread during manual circuit construction, and the advantages of easy removal/representation change completely outweigh the potential concern.
+
+
+## Detailed Design
+
+The new functionality will live in `qiskit.circuit.classical`.
+The expression-builder objects will all be exposed from `qiskit.circuit.classical.expr`, to allow people to bring the desired values into scope either with a namespace prefix `expr` or by a star-import of limited scope.
+
+> From this point on, I will omit the prefix `qiskit.circuit.classical.`.
+
+The class `expr.Expression` will be an abstract base class with two read-only fields:
+```python
+# file: qiskit/circuit/classical/expr.py
+
+class Expression(ABC):
+    resources: typing.Set[Clbit]
+    type: qiskit.circuit.classical.Type
+```
+
+The `resources` field is effectively a `use-def` tracker, which will more easily enable the conversion to data-flow (`DAGCircuit`) format to evaluate the necessary wires of an instruction.
```

Comment:
Not a compiler expert by any means, but I think this is not exactly the same as `use-def` tracking, which makes some assurances about the value of a variable not being re-defined between definition and use. The `Clbit`s in `QuantumCircuit` now follow more like memory semantics, so there are no guarantees of their value or when they were last modified.

Reply:
Yeah, you're right.  In retrospect, I'm not even going to have the `resources` field anyway - I'll just make an `ExpressionVisitor` subclass that calculates them.  There's no reason that these "resources" should be any more special than any other shared property.

Reply:
Fixed in 2f80715.